### PR TITLE
ImmuTable

### PR DESCRIPTION
**Pull request by koke**
I think if you are building a view that's just a table that's really kind of a form, your view controller should only have to specify those cells in a declarative way. The only code there should be how to get a model, turn it into a view model, and handle any changes.

ImmuTable was already introduced in #4510 but I'm extracting it to a separate PR for easier review.

If you want to test it and see it in action, I've reimplemented Settings with ImmuTable in #4549

Needs Review: @jleandroperez

### CODE REVIEW

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+public protocol ImmuTableRow: Reusable {
+    var action: ImmuTableActionType? { get }
+    static var cellClass: AnyClass { get }
+    func configureCell(cell: UITableViewCell)
+    static var registrable: CellRegistrable { get }
+    static var customHeight: Float? { get }
+}
+
+public struct ImmuTableSection {
+    let headerText: String?
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Just wondering if we'd also want support for NSAttributedStrings here?

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    let rows: [ImmuTableRow]
+    let footerText: String?
+
+    init(rows: [ImmuTableRow]) {
+        self.headerText = nil
+        self.rows = rows
+        self.footerText = nil
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Since headerText /  footerText are optionals, no need to initialize them with nil

Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: Optionals don't default to nil, and still need to be initialised

screen shot 2015-12-08 at 14 28 29

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: I think for most cases I can think of, a plain String will be enough. If it becomes necessary we can add it later

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    public static func registerRows(rows: [ImmuTableRow.Type], tableView: CellRegistrator) {
+        let registrables = rows.reduce([:]) {
+            (var classes, row) -> [String: CellRegistrable] in
+
+            classes[row.reusableIdentifier] = row.registrable
+            return classes
+        }
+        for (identifier, registrable) in registrables {
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Just wondering if it wouldn't be the same to just loop thru the *rows* and call *register* ?. (I know, uniques)

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        return viewModel.sections[section].rows.count
+    }
+
+    public func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let row = viewModel.rowAtIndexPath(indexPath)
+        let cell = tableView.dequeueReusableCellWithIdentifier(row.reusableIdentifier, forIndexPath: indexPath)
+
+        row.configureCell(cell)
+
+        configureCell?(cell)
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Two *configureCell* calls sounds a little bit confusing

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+    init(viewModel: ImmuTable) {
+        self.viewModel = viewModel
+    }
+
+    public func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
+        let row = viewModel.rowAtIndexPath(indexPath)
+        if let action = row.action {
+            action(row)
+        }
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Can be replaced with `row.action?(row)` (which looks weird!)

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        let row = viewModel.rowAtIndexPath(indexPath)
+        if let customHeight = row.dynamicType.customHeight {
+            return CGFloat(customHeight)
+        }
+        return tableView.rowHeight
+    }
+}
+
+public struct ImmuTableViewHandler {
+    let target: UITableViewController
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Maybe this one should be weak?

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 3 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: I thought about it for a while. I think it should work, as the docs say

> If you previously registered a class or nib file with the same reuse identifier, the class you specify in the cellClass parameter replaces the old entry. You may specify nil for cellClass if you want to unregister the class from the specified reuse identifier.

However, I feel like that's relying too much on an implementation detail. I just feel safer registering each cell type once.

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 6 above)
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: + How about if you wanna use this in a plain UIViewController?

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            }
+        }
+    }
+
+    lazy var dataSource: ImmuTableDataSource = {
+        return ImmuTableDataSource(viewModel: self.viewModel)
+    }()
+    
+    lazy var delegate: ImmuTableDelegate = {
+        return ImmuTableDelegate(viewModel: self.viewModel)
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: (Suggestion) just to keep things as simple as possible... how about if we merge `ImmuTableDataSource` + `ImmuTableDelegate` into `ImmuTableViewHandler` ?

It'd reduce the footprint, and would probably improve readability.

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    }()
+}
+
+protocol CustomImmuTableRow: ImmuTableRow {
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Same as above, just wondering if we could merge `ImmuTableRow` with the two protocols / extensions below,
and just set as `optional` whatever is not mandatory. It would really improve readability, it's a bit hard to keep track of the whole hierarchy.

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: Huh sorry about that!

**Comment by jleandroperez** [comment]
@koke Code looks great! but it's honestly hard to follow the entire hierarchy. 

Plus, i've got some doubts, such as: 

- If you wanna have (any of the standard cells), with a custom accessory indicator, you'd need to just implement another cell, with the style you want. Just wondering if we wanna extend the `Row` struct to hold more (optional) properties.

- Maybe there's a simpler way (aka less code) to deal with the `MockTableView` concept. (OCMock?). That'd help eliminate a couple protocols.

- Just wondering about the naming consistency detail. `ImmuTable` <> `WPImmuTableCells` (when do we want the *WP* prefix, and when we don't).

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 8 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: >  just set as optional whatever is not mandatory

I tried using optional protocol methods (not sure if here or somewhere else), until I ran into this little issue:

> Optional protocol requirements can only be specified if your protocol is marked with the @objc attribute
> [...]
> Note also that @objc protocols can be adopted only by classes that inherit from Objective-C classes or other @objc classes. They can’t be adopted by structures or enumerations

**Comment by koke** [comment]
> If you wanna have (any of the standard cells), with a custom accessory indicator, you'd need to just implement another cell, with the style you want. Just wondering if we wanna extend the Row struct to hold more (optional) properties.

You'd need to implement another `Row`, as long as your cell isn't a custom class. As for extending Row, remember that `ImmuTableRow` is a protocol, not a base class, so any properties added there would still need to be declared in every type that implements the protocol. We wouldn't have less code, and we'd have many more `nil`s floating around.

> Just wondering about the naming consistency detail. ImmuTable <> WPImmuTableCells (when do we want the WP prefix, and when we don't)

I added the prefix because they inherit from `WPTableViewCell`. I'd love to get rid of that dependency and have `ImmuTableCells` with a sensible default set of cells.

**Comment by koke** [comment]
> Maybe there's a simpler way (aka less code) to deal with the MockTableView concept. (OCMock?). That'd help eliminate a couple protocols.

Maybe there's a _shorter_ way, but I'm not sure if that'd make it better or more readable

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 4 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: Agreed, and I don't think I'm using the data source's configureCell anywhere, so I guess I'll drop it for now.

I think my goal with it was to move the `WPStyleGuide.configureTableViewCell(cell)` out of the custom rows into the view controller, since it's project-specific.

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 6 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: I think it should be `unowned`, and yeah I could probably use a `TableViewController` protocol, but I wonder if that just complicates things, given that we plan to use it with `UITableViewController`

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 7 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: I think I'll explore that option. I built this progressively: initially I just built the data model (ImmuTable), then the data source and delegate, then the UITableViewController subclass, and then rolled that back and implemented a handler instead.

The idea was that it'd be super simple if you needed the default behaviour, but you could go back and use the more specific components if you needed more customisation.

Considering our use cases, I think the best option might be to have the two extremes:
- You use ImmuTableViewHandler, and get all the stuff automatically
- You use ImmuTable as the model, and handle everything else yourself.

File: WordPress/Classes/Utility/ImmuTable.swift (hunk 6 above)
Comment (koke) [inline WordPress/Classes/Utility/ImmuTable.swift]: So this would be easy to implement

https://gist.github.com/koke/b9b7f2e675b6132501bc

Another option would be to make the UITableView the target. Then instead of having to ask the target if the view is loaded, the handler could have a `isViewLoaded` property, or just check if `tableView.superview != nil`. Or maybe we just call reloadData anyway, even if that means calling it twice. To be honest, the current behaviour is tied to UITableViewController:

> When the table view is about to appear the first time it’s loaded, the table-view controller reloads the table view’s data

If we can't be sure that the target VC is a UITableViewController, we can't rely on that behaviour and we should be reloading data when the model is set.

Thoughts?

**Comment by koke** [comment]
@jleandroperez ready for round two, I'm quite happy with this iteration as everything looks much simpler.

I'm going to keep adding some more documentation, but feel free to start with the code review

**Comment by koke** [comment]
I'm done with documentation as well, and reorganised the file to make it more approachable. Please let me know if you find the docs easy to understand.

After this, I'd like to try to move this to a separate library with a bunch of default cells, so it might help with documentation to have a few examples. Also a README separate from the header docs could really help explain what this is and how you're supposed to use it. I added a bit of that at the top of the file, but there's a lot of room for improvement.

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+table view controllers.
+
+It acts as the table view data source and delegate, and signals the table view to
+reload its data when the underlying model changes.
+
+- note: as it keeps a weak reference to its target, you should keep a strong 
+  reference to the handler from your view controller.
+*/
+public class ImmuTableViewHandler: NSObject, UITableViewDataSource, UITableViewDelegate {
+    unowned let target: UITableViewController
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: +1 Looking nice! (Both, unowed, and a single Handler object!)

File: WordPress/Classes/Utility/ImmuTable.swift [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+
+    public func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return viewModel.sections[section].rows.count
+    }
+
+    public func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let row = viewModel.rowAtIndexPath(indexPath)
+        let cell = tableView.dequeueReusableCellWithIdentifier(row.reusableIdentifier, forIndexPath: indexPath)
+
+        row.configureCell(cell)
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/ImmuTable.swift]: This looks great!

**Comment by jleandroperez** [comment]
@koke i really like what you did with ImmuTable!.

I meant to answer this one, but Github doesn't show the Q anymore:

```
extension UITableViewController: ViewControllerWithTable { }
```
Latest implementation looks better than that (less complicated!). There's always room to extend / improve (support for UIViewControllers, ie), but i'm sure ImmuTable can grow as needed / when needed / _if_ needed.

:shipit: !!
